### DISCUSSION TITLE
Using "rofi -show window" with "focus follows mouse" doesn't work well

### DISCUSSION THREAD

**Discussion by novoid** [davatorium/rofi#1585] [Q&A] [answered]
Hi,

I just stumbled over rofi as a suggestion to switch to arbitrary open windows via keyboard shortcut and a search query. Amazing.

### The Problem

However, **`-show window` doesn't work well with "focus follows mouse"** which I prefer to have: switching to a window on a different (virtual) desktop does not give it the focus. Therefore, after using rofi to switch to the window, it is visible but I can't do anything with it without manually moving the mouse. (Please do not remind me of the nature of "focus follows mouse" - I'm well aware of the subtle conflict here. Outside of this issue, "focus follows mouse" does provide me benefits I would not get otherwise.)

I guess I now have identified that **rofi doesn't support focus change** at the moment. I don't know if this qualifies for a feature request. For the moment, I'm settling for a workaround and make it run:

### Potential Workaround

This might be easy to be fixed. I identified `xdotool mousemove -window 0x38000b8 200 200` for moving the mouse relatively to a window ID and I found `rofi -monitor -2 -show window -window-command "wmctrl -i -a {window}"` as a verbose definition of the rofi command where I might re-use the window ID.

Unfortunately, `rofi -monitor -2 -show window -window-command "xdotool mousemove -window {window} 200 200"` does not move the mouse to the window selected by rofi. I don't know what I do wrong here.

My naïve overall-approach which ought to provide me the functionality I need with `rofi -monitor -2 -show window -window-command "wmctrl -i -a {window} ; xdotool mousemove -window {window} 200 200"` does not work for the same reason I guess.

Therefore: **how may I execute two different commands with rofi: one to go to the window and second to move the mouse**?

Thank you for your advice!

**Comment by DaveDavenport** [comment]
> I guess I now have identified that rofi doesn't support focus change at the moment. I don't know if this qualifies for a feature request. For the moment, I'm settling for a workaround and make it run:

I would turn this around: 'switching window with rofi does not work with the focus follow mouse'.
i think it depends also on how focus follow mouse is implemented. (you don't tell use what window manger you use.. so no way for us to check).
Rofi follows the standard (as far as I can tell) to signal via the X server to the window manager to switch focus to the selected window/workspace. 

I think you still solution is with `-window-command`. if you want to execute more then one command, call a shell and execute the multiple commands in there.

**Reply by DaveDavenport** [reply]
In i3 if I turn on `focus_follow_mouse` it works as expected, the focus switches to the window I selected in rofi and not the one under the mouse (until I move the mouse to another window)

**Reply by novoid** [reply]
Yes, I did omit my WM which is xfce (on Xubuntu 20.04 LTS). Sorry.

Do I understand you correctly that rofi is not able to call 2 commands via `-window-command "foo; bara"`?

I tried a wrapper-script as suggested:
```
#!/bin/bash
WINID="${1}"
echo "$(date) ${WINID}" > "${HOME}/log/switch-to-windowid.log"
/usr/bin/wmctrl -i -a "${WINID}"
/usr/bin/xdotool mousemove -window "${WINID}" 200 200
```
It works when called like that: `/home/vk/src/switch-to-windowid.sh 0x38000b8`

Unfortunately, it does not get invoked somehow when called with: `rofi -monitor -2 -show window -window-command "/home/vk/src/switch-to-windowid.sh {window}"` Did I misunderstand something when invoking an external command?

**Reply by DaveDavenport** [reply]
no, rofi executes the command as given.
You can easily call a shell from the command:  `bash -c 'command 1; command2'` lets not re-invent the wheel in rofi.

Did you use the right keybinding? 
copy from manpage:

```markdown
       -window-command cmd

       Set command to execute on selected window for an alt action (-kb-accept-alt).  See PATTERN.

       Default: "wmctrl -i -R {window}"
```

**Reply by novoid** [reply]
> Did you use the right keybinding?

Excuse me, I don't understand the question.

I invoke it manually in the zsh via `rofi -monitor -2 -show window -window-command "/home/vk/src/switch-to-windowid.sh {window}"` and the content of the shellscript is shown above. To my understanding, rofi should invoke `/home/vk/src/switch-to-windowid.sh 0x38000b8` with `0x38000b8` being the selected window ID. Somehow, rofi does not invoke the shell script as I can see in the missing log-file entry.

For debugging this, I prepended the shell command with `G_MESSAGES_DEBUG=all`:

```
[...]

(process:195807): View-DEBUG: 13:53:54.818: expose event
(process:195807): Timings-DEBUG: 13:53:54.818: 6,148198 (0,000106): source/view.c:rofi_view_repaint:234 Expose
(process:195807): Timings-DEBUG: 13:53:54.819: 6,148251 (0,000053): source/view.c:rofi_view_repaint:238 flush

[here, the confirmation of the selection is done]

(process:195807): Rofi-DEBUG: 13:53:55.388: Teardown
(process:195807): View-DEBUG: 13:53:55.388: Cleanup.
(process:195807): View-DEBUG: 13:53:55.390: Unmapping and free'ing window
(process:195807): X11Helper-DEBUG: 13:53:55.391: Cleaning up XCB and XKB
(process:195807): Timings-DEBUG: 13:53:55.391: 6,721031 (0,000000): Stopped
```

There is no log output for the script invocation. However, there is also no invocation log output when I use the default workflow, omitting the `-window-command` entirely.

Therefore, I don't understand what I may be able to check in order to find out why the shell script doesn't get involved. Not only because of the absolute paths I thought I'd be on the safe side.

**Reply by DaveDavenport** [reply]
window-command is only executed when you use the `alt action (-kb-accept-alt)` keybinding. 
To see what key is bound to that, run `rofi -show keys`.

**Reply by novoid** [reply]
Ah, now I see what you mean. Thank you for clarification.

Using Shift-Return (my binding for alt action) **it now works as expected.**

Can you push me into the direction where I may learn how to re-map Return to this alt action so that I don't have to press the Shift modifier every time if this is possible at all?

**Reply by DaveDavenport** [reply]
you can remap every key via the commandline or config file.
Please check the manpages that are shipped with rofi, they exist for a reason!!

https://github.com/davatorium/rofi/blob/next/doc/rofi-keys.5.markdown

**Reply by novoid** [reply]
I understand and thanks for the pointer.

I've been searching the rofi man page and looked at the "SEE ALSO" section where I could find things like "rofi-theme" (thanks for this cool theme subsystem!) but not rofi-keys. With your link from above, I do have everything that **makes the workaround running at my side**: 

`rofi -monitor -2 -show window -kb-accept-alt 'Return' -kb-accept-entry 'Shift+Return' -window-command "/home/vk/src/switch-to-windowid.sh {window}"`

I very much appreciate your time and I'll probably blog about my solution so that others are able to make to most of rofi and xfce when having "focus follows mouse". Greetings from Austria! 🙇

**Reply by DaveDavenport** [reply]
you never told me the version you use, so no way to know what manpages you have.
But the information was (again I don't know your version) in the main manpage for years! under the 'Key Bindings' section (but less verbose).

**Reply by novoid** [reply]
Fair enough: v1.5.4. (via Xubuntu 20.04 LTS) I usually add it to my requests and somehow forgot about that essential information here.

**Reply by DaveDavenport** [reply]
that is very very old.

**Comment by novoid** [comment, marked as answer]
For reference, the root cause of my issue might be this xfce WM bug here: https://gitlab.xfce.org/xfce/xfwm4/-/issues/224

Until this is resolved (the bug is at least 5 years old now), I'll still stick to the workaround discussed here.

**Reply by DaveDavenport** [reply]
Aah, happy to hear it is a known bug and not something unclear. Hopefully they tackle it first.

If answered, can you mark answered the question.

**Reply by novoid** [reply]
I agree, that this is no bug of rofi with respect to that.

Since the xfce bug is at leat 5 years old, I don't think that they're going to fix it soon. Therefore, I would still like to make the mitigating workaround run where an external shell script should be invoked by rofi.

**Comment by 42sol-eu** [comment]
I used the ` -normal-window` parameter to get rofi closer to the use case I wanted it.
I think this is a nice middle way still using the focus-follow-mouse and get the focus to the (much bigger) rofi window
